File: flutter_skyway_ios/ios/Classes/FlutterSkywayPlugin.swift
import Flutter
import UIKit
import SkyWayRoom

enum ChannelName {
    static let skywayMethodChannel = "flutter_skyway_ios"
    static let joinCallEventChannel = "flutter_skyway_ios/join_call_channel"
    static let memberListChangeEventChannel = "flutter_skyway_ios/member_list_change_channel"
    static let memberJoinedEventChannel = "flutter_skyway_ios/member_joined_channel"
    static let memberLeftEventChannel = "flutter_skyway_ios/member_left_channel"
    static let localRendererPlatformView = "flutter_skyway_ios/local_renderer"
    static let remoteRendererPlatformView = "flutter_skyway_ios/remote_renderer"
}

public class FlutterSkywayPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, RoomDelegate, RemoteDataStreamDelegate  {
    
    private var room: SFURoom?
    private var localRoomMember: LocalRoomMember?
    private var localVideoStream: LocalVideoStream?
    private var localAudioStream: LocalAudioStream?
    private var flutterPluginRegistrar: FlutterPluginRegistrar?
    
    private var  joinCallEventSink: FlutterEventSink?
    private var memberListChangeEventSink: FlutterEventSink?
    private var memberJoinedEventSink: FlutterEventSink?
    private var memberLeftEventSink: FlutterEventSink?
    
    private var localView: UIView?
    
    private var remoteRoomMemberList:[RemoteUIView] = []
    
    private var audioPublication: RoomPublication?
    private var videoPublication: RoomPublication?
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: ChannelName.skywayMethodChannel, binaryMessenger: registrar.messenger())
        let instance = FlutterSkywayPlugin()
        instance.flutterPluginRegistrar = registrar
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.registerEventChannel(registrar)
        instance.registerViewFactory(registrar)
    }
    
    public func registerEventChannel(_ registrar: FlutterPluginRegistrar) {
        // Join call event channel
        let joinCallEventChannel = FlutterEventChannel(name: ChannelName.joinCallEventChannel, binaryMessenger: registrar.messenger())
        joinCallEventChannel.setStreamHandler(self)
        
        let memberListChangeEventChannel = FlutterEventChannel(name: ChannelName.memberListChangeEventChannel, binaryMessenger: registrar.messenger())
        memberListChangeEventChannel.setStreamHandler(self)
        
        let memberJoinedEventChannel = FlutterEventChannel(name: ChannelName.memberJoinedEventChannel, binaryMessenger: registrar.messenger())
        memberJoinedEventChannel.setStreamHandler(self)
        
        let memberLeftEventChannel = FlutterEventChannel(name: ChannelName.memberLeftEventChannel, binaryMessenger: registrar.messenger())
        memberLeftEventChannel.setStreamHandler(self)
    }
    
    public func registerViewFactory(_ registrar: FlutterPluginRegistrar) {
        localView = UIView()
        registrar.register(
            FlutterSkywayPlatformViewFactory(messenger: registrar.messenger(), view: localView!), withId: ChannelName.localRendererPlatformView
        )
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch(call.method) {
        case "connect":
            onConnect(call: call, result: result)
            break
        case "joinRoom":
            joinRoom(call: call, result: result)
            break
        case "leaveRoom":
            leaveRoom(call: call, result: result)
            break
        case "getRemoteMemberList":
            getRemoteMemberList(call: call, result: result)
            break
        case  "getLocalMember" :
            getLocalMember(call: call, result: result)
            break
        case  "toggleMic" :
            toggleMic(call: call, result: result)
            break
        default:
            break
        }
    }
    
    public func onConnect(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String:Any]
        if let args = arguments {
            let authToken = args["authToken"] as? String ?? ""
            Task {
                do {
                    let opt: ContextOptions = .init()
                        opt.logLevel = .trace
                    try await Context.setup(withToken: authToken, options: opt)
                    result("success")
                }
                catch {
                    result("fail")
                }
            }
            
        } else {
            result("fail")
        }
    }
    
    public func joinRoom(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String:Any]
        if let args = arguments {
            let memberName = args["memberName"] as? String ?? ""
            let roomName = args["roomName"] as? String ?? ""
            let isVideo = args["isVideo"] as? Bool ?? false
            Task {
                DispatchQueue.main.sync {
                    self.joinCallEventSink?("joining")
                }
                // Create room
                let roomInit: Room.InitOptions = .init()
                roomInit.name = roomName
                room = try? await SFURoom.findOrCreate(with: roomInit)
                if(room == nil){
                    result(nil)
                    return
                }
                room?.delegate = self
                // Join room
                let memberInit: Room.MemberInitOptions = .init()
                memberInit.name = memberName
                localRoomMember = try? await room?.join(with: memberInit)
                if(localRoomMember == nil){
                    result(nil)
                    DispatchQueue.main.sync {
                        self.joinCallEventSink?("joinFail")
                    }
                    return
                }
                fetchRemoteMemberList()
                
                // AudioStreamの作成
                let audioSource: MicrophoneAudioSource = .init()
                localAudioStream = audioSource.createStream()

                // Publish local audio, video
                let pubAudio = try? await localRoomMember?.publish(localAudioStream!, options: nil)
                audioPublication = pubAudio
                if( pubAudio == nil){
                    print("Publishing audio failed.")
                }

                if(isVideo){
                // Cameraの設定
                    let camera = CameraVideoSource.supportedCameras().first(where: { $0.position == .front })

                    if(camera != nil) {
                        try! await CameraVideoSource.shared().startCapturing(with: camera!, options: nil)
                    }else{
                        print("Supported cameras is not found.");
                    }

                    // VideoStreamの作成
                    localVideoStream = CameraVideoSource.shared().createStream()
                    let cameraView = await CameraPreviewView()
                    // Previewの描画
                    DispatchQueue.main.sync {
                        cameraView.translatesAutoresizingMaskIntoConstraints = false
                        cameraView.frame =
                        CGRect(
                            x: 0, y: 0,
                            width: 200,
                            height: 200
                        )
                        cameraView.contentMode = .scaleAspectFit
                        CameraVideoSource.shared().attach(cameraView)
                        localView?.addSubview(cameraView)

                        NSLayoutConstraint.activate([
                            cameraView.leadingAnchor.constraint(equalTo: localView!.leadingAnchor),
                            cameraView.trailingAnchor.constraint(equalTo: localView!.trailingAnchor),
                            cameraView.topAnchor.constraint(equalTo: localView!.topAnchor),
                            cameraView.bottomAnchor.constraint(equalTo: localView!.bottomAnchor)
                        ])
                    }

                    let pubVideo = try? await localRoomMember?.publish(localVideoStream!, options: nil)
                    videoPublication = pubVideo
                    if(pubVideo == nil){
                        print(" Publishing video failed.")
                    }
                }

                // Subscribe remote publication
                await subscribePublications()


                // Join room success
                result(roomMemberString(member: localRoomMember!))
                DispatchQueue.main.sync {
                    self.joinCallEventSink?("joinSuccess")
                }
            }
            
        } else {
            result(nil)
            DispatchQueue.main.sync {
                self.joinCallEventSink?("joinFail")
            }
        }
    }
    
    public func leaveRoom(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            if (localRoomMember != nil) {
                try? await localRoomMember?.leave()
                try? await room?.leave(localRoomMember!)
                try? await room?.dispose()
                localRoomMember = nil
                room = nil
                audioPublication = nil
                videoPublication = nil
                for subview in await localView!.subviews {
                    await subview.removeFromSuperview()
                }
                CameraVideoSource.shared().stopCapturing()
                try? await Context.dispose()
                result("success")
            }else{
                result(nil)
            }
        }
    }
    
    public func getRemoteMemberList(call: FlutterMethodCall, result: @escaping FlutterResult){
        fetchRemoteMemberList()
        var memberJsonList = [String]()
        for member in remoteRoomMemberList {
            if let data = roomMemberInfoString(member: member.member){
                memberJsonList.append(data)
            }
        }
        result("\(memberJsonList)")
    }
    
    private func getLocalMember(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if (localRoomMember != nil) {
            result(roomMemberString(member: localRoomMember!))
        } else {
            result(nil)
        }
    }
    
    public func fetchRemoteMemberList() {
        let members:[RoomMember] = room?.members ?? []
        for (remoteIndex, remoteMem) in remoteRoomMemberList.enumerated() {
            let isInRoom = members.firstIndex(where: { $0.id == remoteMem.member.id }) != -1
            if (!isInRoom) {
                remoteRoomMemberList.remove(at: remoteIndex)
            }
        }
        for member in members {
            addRemoteMemberAndRegisterView(member: member)
        }
    }
    
    private func addRemoteMemberAndRegisterView(
        member: RoomMember,
        completion: (() -> Void)? = nil
    ) {
        let isRemoteMemberInList = remoteRoomMemberList.contains { $0.member.id == member.id }
        if(isRemoteMemberInList){
            return
        }
        if(member.id == localRoomMember?.id || localRoomMember == nil){
            return
        }
        DispatchQueue.main.async {
            let remoteView = UIView()
            self.flutterPluginRegistrar!.register(
                FlutterSkywayPlatformViewFactory(messenger: self.flutterPluginRegistrar!.messenger(), view: remoteView), withId:  "\(ChannelName.remoteRendererPlatformView)/\(member.id)"
            )
            self.remoteRoomMemberList.append(
                RemoteUIView(
                    member: RemoteRoomMemberInfo(id: member.id, name: member.name),
                    uiView:  remoteView
                )
            )
        }
        if(completion != nil){
            completion!()
        }
    }
    
    public func subscribePublications() async {
        for pub in room?.publications ?? [] {
            if(pub.publisher?.id != localRoomMember?.id){
                subscribe(publication: pub)
            }
        }
    }
    
    private func toggleMic(call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            let arguments = call.arguments as? [String:Any]
            if let args = arguments {
                let isMute = args["isMute"] as? Bool ?? false
                if (isMute) {
                    try? await audioPublication?.disable()
                } else {
                    try? await audioPublication?.enable()
                }
                result("success")
            }
        }
    }
    
    private func unPublish(roomPublication: RoomPublication) {
        Task{
            try? await localRoomMember?.unpublish(publicationId: roomPublication.id)
        }
    }
    
    // MARK: - RoomDelegate
    public func room(_ room: Room, didPublishStreamOf publication: RoomPublication) {
        if publication.publisher?.id != localRoomMember?.id {
            subscribe(publication: publication)
        }
    }
    
    public func roomMemberListDidChange(_ room: Room) {
        var memberJsonList = [String]()
        for member in self.room?.members ?? [] {
            if let data = roomMemberString(member: member){
                memberJsonList.append(data)
            }
        }
        print("roomMemberListDidChange: \(memberJsonList)")
        DispatchQueue.main.sync {
            self.memberListChangeEventSink?("\(memberJsonList)")
        }
    }
    
    public func room(_ room: Room, memberDidJoin member: RoomMember) {
        DispatchQueue.main.sync {
            addRemoteMemberAndRegisterView(member: member) { [self] in
                self.memberJoinedEventSink?(self.roomMemberString(member: member))
            }
        }
    }
    
    public func room(_ room: Room, memberDidLeave member: RoomMember) {
        DispatchQueue.main.sync {
            let index = remoteRoomMemberList.firstIndex(where: { $0.member.id == member.id })
            if(index == -1 || index == nil){
                return
            }
            self.memberLeftEventSink?(roomMemberString(member: member))
            remoteRoomMemberList[index!].uiView = UIView()
            remoteRoomMemberList.remove(at: index!)
        }
    }
    
    public func subscribe(publication: RoomPublication) {
        Task {
            let opt: SubscriptionOptions = .init()
            let sub = try? await localRoomMember?.subscribe(publicationId: publication.id, options: opt)
            if(sub == nil){
                return
            }
            DispatchQueue.main.sync {
                if(sub?.stream is RemoteVideoStream){
                    let index =
                    remoteRoomMemberList.firstIndex(where: { $0.member.id == publication.publisher?.id })
                    if (index == -1 || index == nil){
                        return
                    }
                    let videoView = VideoView()
                    videoView.frame =
                    CGRect(
                        x: 0, y: 0,
                        width: 200,
                        height: 200
                    )
                    videoView.videoContentMode = .scaleAspectFit
                    let remoteVideoStream = sub!.stream as! RemoteVideoStream
                    remoteVideoStream.attach(videoView)
                    remoteRoomMemberList[index!].uiView.addSubview(videoView)
                }
            }
        }
    }
    
    public func roomMemberString(member: RoomMember)-> String?{
        let data: [String: Any] = [
            "memberId": member.id,
            "memberName": member.name ?? "",
            "isRemoteMember": member.id != localRoomMember?.id,
            "isMute" : member.id == localRoomMember?.id ? audioPublication == nil : false,
            "isVideo": member.id == localRoomMember?.id ? videoPublication != nil : false
        ]
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: data, options: [])
            if let jsonString = String(data: jsonData, encoding: .utf8) {
                return jsonString
            }
        } catch {
            print("Error: \(error)")
        }
        return nil
    }
    
    func roomMemberInfoString(member: RemoteRoomMemberInfo)-> String?{
        let data: [String: Any] = [
            "memberId": member.id,
            "memberName": member.name ?? "",
            "isRemoteMember": member.id != localRoomMember?.id,
            "isMute" : member.id == localRoomMember?.id ? audioPublication == nil : false,
            "isVideo": member.id == localRoomMember?.id ? videoPublication != nil : false
        ]
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: data, options: [])
            if let jsonString = String(data: jsonData, encoding: .utf8) {
                return jsonString
            }
        } catch {
            print("Error: \(error)")
        }
        return nil
    }
    
    // Assign event channel sink
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        
        if arguments as? String == ChannelName.joinCallEventChannel {
            joinCallEventSink = events
        }
        
        if arguments as? String == ChannelName.memberListChangeEventChannel {
            memberListChangeEventSink = events
        }
        if arguments as? String == ChannelName.memberJoinedEventChannel {
            memberJoinedEventSink = events
        }
        
        if arguments as? String == ChannelName.memberLeftEventChannel {
            memberLeftEventSink = events
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        joinCallEventSink = nil
        return nil
    }
    
}

class RemoteUIView {
    var member: RemoteRoomMemberInfo
    var uiView: UIView
    
    init(member: RemoteRoomMemberInfo, uiView: UIView) {
        self.member = member
        self.uiView = uiView
    }
}

class RemoteRoomMemberInfo {
    var id: String
    var name: String?
    
    init(id: String, name: String?) {
        self.id = id
        self.name = name
    }
}

File: flutter_skyway_android/android/src/main/kotlin/com/vuitv/flutter_skyway/FlutterSkywayPlugin.kt
package com.vuitv.flutter_skyway

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import androidx.annotation.NonNull
import androidx.constraintlayout.widget.ConstraintLayout
import com.ntt.skyway.core.SkyWayContext
import com.ntt.skyway.core.content.Stream
import com.ntt.skyway.core.content.local.LocalAudioStream
import com.ntt.skyway.core.content.local.LocalVideoStream
import com.ntt.skyway.core.content.local.source.AudioSource
import com.ntt.skyway.core.content.local.source.CameraSource
import com.ntt.skyway.core.content.remote.RemoteAudioStream
import com.ntt.skyway.core.content.remote.RemoteDataStream
import com.ntt.skyway.core.content.remote.RemoteVideoStream
import com.ntt.skyway.core.content.sink.SurfaceViewRenderer
import com.ntt.skyway.core.util.Logger
import com.ntt.skyway.room.RoomPublication
import com.ntt.skyway.room.RoomSubscription
import com.ntt.skyway.room.member.LocalRoomMember
import com.ntt.skyway.room.member.RemoteRoomMember
import com.ntt.skyway.room.member.RoomMember
import com.ntt.skyway.room.sfu.SFURoom

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.json.JSONObject

class FlutterSkywayPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var context: Context? = null

    //
    private lateinit var option: SkyWayContext.Options

    // メンバの宣言
    private val scope = CoroutineScope(Dispatchers.IO)
    private var localRoomMember: LocalRoomMember? = null
    private var room: SFURoom? = null
    private var localVideoStream: LocalVideoStream? = null
    private var localAudioStream: LocalAudioStream? = null

    private var audioPublication: RoomPublication? = null
    private var videoPublication: RoomPublication? = null

    private lateinit var localConstraintLayout: ConstraintLayout

    private val joinCallEventChannel = "flutter_skyway_android/join_call_channel"
    private var joinCallEventSink: EventChannel.EventSink? = null

    private val memberListChangeEventChannel = "flutter_skyway_android/member_list_change_channel"
    private var memberListChangeEventSink: EventChannel.EventSink? = null

    private var memberJoinedEventSink: EventChannel.EventSink? = null
    private val memberJoinedEventChannel = "flutter_skyway_android/member_joined_channel"

    private var memberLeftEventSink: EventChannel.EventSink? = null
    private val memberLeftEventChannel = "flutter_skyway_android/member_left_channel"

    private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding

    private var remoteRoomMemberList: MutableList<RemoteConstrainView> = mutableListOf()


    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_skyway_android")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        this.flutterPluginBinding = flutterPluginBinding;
        localConstraintLayout = LayoutInflater.from(context).inflate(
            R.layout.local_render_layout, null
        ) as ConstraintLayout

        registerEventChannel()
        registerViewFactory()
    }


    private fun registerEventChannel() {
        /// Event channel
        EventChannel(flutterPluginBinding.binaryMessenger, joinCallEventChannel).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(p0: Any?, eventSink: EventChannel.EventSink) {
                    joinCallEventSink = eventSink
                }

                override fun onCancel(p0: Any) {
                    joinCallEventSink = null
                }
            },
        )

        EventChannel(
            flutterPluginBinding.binaryMessenger, memberListChangeEventChannel
        ).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(p0: Any?, eventSink: EventChannel.EventSink) {
                    memberListChangeEventSink = eventSink
                }

                override fun onCancel(p0: Any) {
                    memberListChangeEventSink = null
                }
            },
        )

        EventChannel(
            flutterPluginBinding.binaryMessenger, memberJoinedEventChannel
        ).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(p0: Any?, eventSink: EventChannel.EventSink) {
                    memberJoinedEventSink = eventSink
                }

                override fun onCancel(p0: Any) {
                    memberJoinedEventSink = null
                }
            },
        )

        EventChannel(
            flutterPluginBinding.binaryMessenger, memberLeftEventChannel
        ).setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(p0: Any?, eventSink: EventChannel.EventSink) {
                    memberLeftEventSink = eventSink
                }

                override fun onCancel(p0: Any) {
                    memberLeftEventSink = null
                }
            },
        )
    }

    private fun registerViewFactory() {
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "flutter_skyway_android/local_renderer",
            FlutterSkywayPlatformViewFactory(localConstraintLayout),
        )
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformName" -> {
                result.success("Android SkyWay")
            }

            "connect" -> {
                onConnect(call, result)
            }

            "joinRoom" -> {
                onJoinRoom(call, result)
            }

            "leaveRoom" -> {
                leaveRoom(call, result)
            }

            "subscribePublications" -> {
                subscribePublications()
            }

            "getRemoteMemberList" -> {
                getRoomRemoteMemberList(call, result)
            }

            "getLocalMember" -> {
                getLocalMember(call, result)
            }

            "toggleMic" -> {
                toggleMic(call, result)
            }

            else -> result.notImplemented()
        }

    }


    private fun onConnect(@NonNull call: MethodCall, @NonNull result: Result) {
        val authToken = call.arguments<HashMap<String, String>>()?.get("authToken") ?: return
        option = SkyWayContext.Options(
            authToken = authToken, logLevel = Logger.LogLevel.VERBOSE, enableHardwareCodec = false
        )

        scope.launch() {
            val connect = context?.let {
                SkyWayContext.setup(it, option)
            }
            GlobalScope.launch(Dispatchers.Main) {
                if (connect == false) {
                    result.success("fail")
                } else {
                    result.success("success")
                }
            }

        }
    }

    private fun onJoinRoom(@NonNull call: MethodCall, @NonNull result: Result) {
        val memberName = call.arguments<HashMap<String, String>>()?.get("memberName") ?: return
        val roomName = call.arguments<HashMap<String, String>>()?.get("roomName") ?: return
        val isVideo = call.arguments<HashMap<String, Boolean>>()?.get("isVideo") ?: return

        scope.launch() {
            GlobalScope.launch(Dispatchers.Main) {
                joinCallEventSink?.success("joining")
            }
            room = SFURoom.findOrCreate(roomName)

            if (room == null) {
                GlobalScope.launch(Dispatchers.Main) {
                    result.success(null)
                    joinCallEventSink?.success("joinFail");
                }
                return@launch
            }

            val memberInit = RoomMember.Init(name = memberName)
            localRoomMember = room?.join(memberInit)

            val resultMessage = if (localRoomMember == null) "Join failed" else "Joined room"
            Log.d("FlutterSkywayPlugin - resultMessage", resultMessage)
            if (localRoomMember == null) {
                GlobalScope.launch(Dispatchers.Main) {
                    result.success(null)
                    joinCallEventSink?.success("joinFail");
                }
                return@launch
            }
            fetchRemoteMemberList()

            // 音声入力を開始します
            AudioSource.start()
            localAudioStream = AudioSource.createStream()
            audioPublication = localRoomMember?.publish(localAudioStream!!)

            if (isVideo) {
                // cameraリソースの取得
                val device = context?.let { CameraSource.getFrontCameras(it).first() }
                // camera映像のキャプチャを開始します
                val cameraOption = CameraSource.CapturingOptions(800, 800)
                context?.let {
                    if (device != null) {
                        CameraSource.startCapturing(it, device, cameraOption)
                    }
                }
                // 描画やpublishが可能なStreamを作成します
                localVideoStream = CameraSource.createStream()

                GlobalScope.launch(Dispatchers.Main) {
                    val localViewRenderer = context?.let { SurfaceViewRenderer(it) }!!
                    localConstraintLayout.removeAllViews()
                    localConstraintLayout.addView(localViewRenderer)
                    localViewRenderer.setup()
                    localVideoStream?.addRenderer(localViewRenderer)
                }
                // publish local
                videoPublication = localRoomMember?.publish(localVideoStream!!)
            }

            subscribePublications()
            roomHandler()

            GlobalScope.launch(Dispatchers.Main) {
                joinCallEventSink?.success("joinSuccess")
                result.success(roomMemberJson(localRoomMember!!).toString())
            }
        }
    }

    private fun subscribePublications() {
        // 入室時に他のメンバーのStreamを購読する
        room?.publications?.forEach {
            if (it.publisher?.id == localRoomMember?.id) {
                return@forEach
            }
            subscribe(it)
        }
        // 誰かがStreamを公開したときに購読する
        room?.onStreamPublishedHandler = Any@{
            Log.d("room", "onStreamPublished: ${it.id}")
            if (it.publisher?.id == localRoomMember?.id) {
                return@Any
            }
            subscribe(it)
        }
    }

    private fun roomHandler() {
        room?.onMemberListChangedHandler = {
            GlobalScope.launch(Dispatchers.Main) {
                memberListChangeEventSink?.success(
                    "${
                        room?.members?.map { roomMemberJson(it) }?.toMutableList() ?: listOf()
                    }"
                )
            }
        }
        room?.onMemberJoinedHandler = {
            GlobalScope.launch(Dispatchers.Main) {
                addRemoteMemberAndRegisterView(it) {
                    memberJoinedEventSink?.success("${roomMemberJson(it)}")
                }
            }
        }
        room?.onMemberLeftHandler = {
            GlobalScope.launch(Dispatchers.Main) {
                val index = remoteRoomMemberList.indexOfFirst { it1 -> it1.member.id == it.id }
                if (index == -1) return@launch
                memberLeftEventSink?.success("${roomMemberJson(it)}")
                remoteRoomMemberList[index].constraintLayout.removeAllViews()
                remoteRoomMemberList.removeAt(index)
            }
        }
    }

    private fun getRoomRemoteMemberList(@NonNull call: MethodCall, @NonNull result: Result) {
        fetchRemoteMemberList()
        result.success(
            "${
                remoteRoomMemberList.map { roomMemberJson(it.member) }.toMutableList()
            }"
        )
    }

    private fun getLocalMember(@NonNull call: MethodCall, @NonNull result: Result) {
        if (localRoomMember != null) {
            result.success(
                "${
                    roomMemberJson(localRoomMember!!)
                }"
            )
        } else {
            result.success(null)
        }
    }

    private fun toggleMic(@NonNull call: MethodCall, @NonNull result: Result) {
        scope.launch() {
            val isMute = call.arguments<HashMap<String, Boolean>>()?.get("isMute") ?: false
            scope.launch(Dispatchers.Main) {
                if (isMute) {
                    audioPublication?.disable()
//                    localAudioStream?.track?.setEnabled(false)
                } else {
                    audioPublication?.enable()
//                    localAudioStream?.track?.setEnabled(true)
                }
            }
            result.success("success")
        }
    }

    private fun unPublish(roomPublication: RoomPublication) {
        scope.launch(Dispatchers.IO) {
            localRoomMember?.unpublish(roomPublication)
        }
    }

    private fun fetchRemoteMemberList() {
        val members: List<RoomMember> = (room?.members ?: listOf()).toMutableList()
        remoteRoomMemberList.forEach(action = {
            val isInRoom = members.indexOfFirst { it1 -> it1.id == it.member.id } != -1
            if (!isInRoom) {
                remoteRoomMemberList.remove(it)
            }
        })
        members.forEach(
            action = {
                addRemoteMemberAndRegisterView(it)
            },
        )
    }

    private fun addRemoteMemberAndRegisterView(
        member: RoomMember, callback: (() -> Unit)? = null
    ) {
        var isRemoteMemberInList = remoteRoomMemberList.any { it.member.id == member.id }
        if (isRemoteMemberInList) {
            return
        }
        if (member.id == localRoomMember?.id || localRoomMember == null) {
            return
        }
        var remoteConstraintLayout = LayoutInflater.from(context)
            .inflate(R.layout.remote_render_layout, null) as ConstraintLayout
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "flutter_skyway_android/remote_renderer/${member.id}",
            FlutterSkywayPlatformViewFactory(remoteConstraintLayout),
        )
        remoteRoomMemberList.add(
            RemoteConstrainView(
                member as RemoteRoomMember, remoteConstraintLayout
            )
        )
        if (callback == null) return
        callback.invoke()
    }

    private fun roomMemberJson(it: RoomMember): JSONObject {
        val rootObject = JSONObject()
        rootObject.put("memberId", it.id)
        rootObject.put("memberName", it.name)
        rootObject.put("isRemoteMember", it.id != localRoomMember?.id)
        if (it.id == localRoomMember?.id) {
            rootObject.put("isMute", audioPublication == null)
            rootObject.put("isVideo", videoPublication != null)
        }
        rootObject
        return rootObject
    }


    // 購読(subscribe)の実装部分
    private fun subscribe(publication: RoomPublication) {
        scope.launch {
            // Publicationをsubscribeします
            val subscription: RoomSubscription? = localRoomMember?.subscribe(publication.id)

            GlobalScope.launch(Dispatchers.Main) {
                val remoteStream = subscription?.stream ?: return@launch

                when (remoteStream.contentType) {
                    // コンポーネントへの描画
                    Stream.ContentType.VIDEO -> {
                        val index =
                            remoteRoomMemberList.indexOfFirst { it1 -> it1.member.id == publication.publisher?.id }
                        if (index == -1) return@launch
                        remoteRoomMemberList[index].constraintLayout.removeAllViews()
                        val remoteViewRenderer = context?.let { SurfaceViewRenderer(it) }!!
                        remoteRoomMemberList[index].constraintLayout.addView(remoteViewRenderer)
                        remoteViewRenderer.setup()
                        (remoteStream as RemoteVideoStream).addRenderer(
                            remoteViewRenderer
                        )
                    }

                    Stream.ContentType.AUDIO -> {
                        (subscription.stream as RemoteAudioStream)
                    }

                    Stream.ContentType.DATA -> {
                        (subscription.stream as RemoteDataStream).onDataHandler = {}

                        (subscription.stream as RemoteDataStream).onDataBufferHandler = {}
                    }

                    else -> {}
                }
            }
        }
    }

    private fun leaveRoom(@NonNull call: MethodCall, @NonNull result: Result) {
        if (localRoomMember != null && SkyWayContext.isSetup) {
            scope.launch(Dispatchers.Main) {
                room?.leave(localRoomMember!!)
                localRoomMember?.leave()
                room?.dispose()
                room = null
                localRoomMember = null
                localAudioStream = null
                audioPublication = null
                videoPublication = null
                localVideoStream?.removeAllRenderer()
                CameraSource.stopCapturing()
                SkyWayContext.dispose()
                result.success("success")
            }
        } else {
            result.success(null)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        if (localRoomMember != null) {
            scope.launch(Dispatchers.IO) {
                room?.leave(localRoomMember!!)

            }
        }
        scope.launch(Dispatchers.IO) {
            room?.dispose()
            room = null
            channel.setMethodCallHandler(null)
            context = null
            CameraSource.stopCapturing()
            audioPublication = null
            videoPublication = null
            localRoomMember = null
            localAudioStream = null
            remoteRoomMemberList.clear()
            SkyWayContext.dispose()
        }
        GlobalScope.launch(Dispatchers.Main) {
            joinCallEventSink?.success("leftRoom")
        }

    }
}

class RemoteConstrainView(member: RemoteRoomMember, constraintLayout: ConstraintLayout) {
    var member: RemoteRoomMember = member
    var constraintLayout: ConstraintLayout = constraintLayout
}